go/externals/services.go

```diff
@@ -4,37 +4,86 @@
 package externals
 
 import (
+	"fmt"
 	"strings"
+	"sync"
 
 	libkb "github.com/keybase/client/go/libkb"
 )
 
-type externalServicesCollection map[string]libkb.ServiceType
+type externalServices struct {
+	sync.Mutex
+	libkb.Contextified
+	collection map[string]libkb.ServiceType
+	loaded     bool
+}
+
+func NewExternalServices(g *libkb.GlobalContext) libkb.ExternalServicesCollector {
+	e := &externalServices{
+		Contextified: libkb.NewContextified(g),
+		collection:   make(map[string]libkb.ServiceType),
+	}
 
-var externalServices = externalServicesCollection(make(map[string]libkb.ServiceType))
+	nonGenericServiceTypes := []libkb.ServiceType{
+		DNSServiceType{},
+		FacebookServiceType{},
+		GithubServiceType{},
+		HackerNewsServiceType{},
+		RedditServiceType{},
+		RooterServiceType{},
+		TwitterServiceType{},
+		WebServiceType{},
+	}
+	e.Lock()
+	defer e.Unlock()
+	for _, st := range nonGenericServiceTypes {
+		e.register(st)
+	}
+	return e
+}
 
-func (e externalServicesCollection) Register(st libkb.ServiceType) {
+func (e *externalServices) register(st libkb.ServiceType) {
 	for _, k := range st.AllStringKeys() {
-		e[k] = st
+		e.collection[k] = st
 	}
 }
 
-func (e externalServicesCollection) GetServiceType(s string) libkb.ServiceType {
-	return e[strings.ToLower(s)]
+func (e *externalServices) GetServiceType(s string) libkb.ServiceType {
+	e.Lock()
+	defer e.Unlock()
+	e.loadGenericServices()
+	return e.collection[strings.ToLower(s)]
 }
 
-func (e externalServicesCollection) ListProofCheckers(mode libkb.RunMode) []string {
+func (e *externalServices) ListProofCheckers(mode libkb.RunMode) []string {
+	e.Lock()
+	defer e.Unlock()
+	e.loadGenericServices()
 	var ret []string
-	for k, v := range e {
+	for k, v := range e.collection {
 		if useDevelProofCheckers || !v.IsDevelOnly() {
 			ret = append(ret, k)
 		}
 	}
 	return ret
 }
 
-var _ libkb.ExternalServicesCollector = externalServices
+func (e *externalServices) loadGenericServices() {
+	// TODO need to hook this into storage for generic configs
+	shouldRun := e.G().Env.GetFeatureFlags().Admin() || e.G().Env.GetRunMode() == libkb.DevelRunMode || e.G().Env.RunningInCI()
+
+	if !shouldRun || e.loaded {
+		return
+	}
 
-func GetServices() libkb.ExternalServicesCollector {
-	return externalServices
+	loader := NullConfigLoader{}
+	configs, err := loader.LoadAll()
+	if err != nil {
+		// TODO we probably don't want it to work this way..
+		panic(fmt.Sprintf("Unable to load proofs: %v", err))
```

Comment:
heh yeah. Panic makes me nervous. Log and proceed as if configs was empty? Add ticket number to TODO